include/lsst/afw/image/Pixel.h

```diff
@@ -652,4 +661,28 @@ std::ostream& operator<<(std::ostream& os, BinaryExpr<ExprT1, ExprT2, BinOp, Mas
 }  // namespace image
 }  // namespace afw
 }  // namespace lsst
+
+namespace std {
+template <typename ImagePixelT, typename MaskPixelT, typename VariancePixelT>
+struct hash<lsst::afw::image::pixel::Pixel<ImagePixelT, MaskPixelT, VariancePixelT>> {
+    using argument_type = lsst::afw::image::pixel::Pixel<ImagePixelT, MaskPixelT, VariancePixelT>;
+    using result_type = size_t;
+    size_t operator()(argument_type const& obj) const noexcept { return obj.hash_value(); }
+};
+
+template <typename ImagePixelT, typename MaskPixelT, typename VariancePixelT>
+struct hash<lsst::afw::image::pixel::SinglePixel<ImagePixelT, MaskPixelT, VariancePixelT>> {
+    using argument_type = lsst::afw::image::pixel::SinglePixel<ImagePixelT, MaskPixelT, VariancePixelT>;
+    using result_type = size_t;
+    size_t operator()(argument_type const& obj) const noexcept {
+        // Don't want to define SinglePixel::hash_value when SinglePixel::operator== doesn't exist
```

Comment:
In that case why are we defining `std::hash<SinglePixel<...>>`?  Is it used in hashing some other class somewhere I didn't spot?

Reply:
Ugh, yes, the relationship between `Pixel` and `SinglePixel`... 😡

The problem (or, at least, the relevant one) is that `SinglePixel` *can* participate in equality tests -- but only on the right-hand side of `operator==(Pixel, T)`. And given all the hacking that went into making `Pixel` and `SinglePixel` interchangeable (e.g., [unsafe implicit conversions](https://github.com/lsst/afw/blob/tickets/DM-9938/include/lsst/afw/image/Pixel.h#L159-L165)), I wouldn't be surprised if there were some code *somewhere* that tries to mix them in ways where they might end up being hashed at the same time.